src/chunks/vocab/mod.rs

```diff
@@ -65,13 +67,15 @@ impl WordIndex {
     }
 }
 
-pub(crate) fn create_indices(words: &[String]) -> HashMap<String, usize> {
-    let mut indices = HashMap::new();
-
+pub(crate) fn create_indices<'a, 'b>(words: &'b [String]) -> HashMap<&'a str, usize> {
```

Comment:
In order for this to be safe, the lifetime of the strings in the `HashMap` should be tied to that of the strings in `words`, otherwise, it would be possible to deallocate the strings in `words` and the pointers in the hashmap would be wild pointers. In this implementation, it is definitely unsafe, because the strings in the `HashMap` have a static lifetime.

Reply:
I know that there's a mismatch but since `index` is only used internally by the vocabularies and cannot be accessed in any way from the outside, the fake-static lifetime shouldn't cause any issues, right? Or does this cause the pointers not to be dropped when the vocab (and index) go out of scope?

Do you see another way or do you know of a crate that does this properly?

Reply:
> I know that there's a mismatch but since `index` is only used internally by the vocabularies and cannot be accessed in any way from the outside, the fake-static lifetime shouldn't cause any issues, right?

It does cause issues. Clone the vocabulary, let the original go out of scope, your cloned vocab has dead pointers (to the original `String`s. So, you have to implement Clone yourself. And that's the problem, once you remove this safety, with every API change you have to take into account whether there is a situation where it could result in wild pointers. These are exactly the kinds of issues that the Rust ownership system prevents.

If we want the memory savings (for regular vocabularies it's not so pressing), use proper sharing (`Rc`, but we cannot do that since it is not `Send + Sync`, and `Arc` is probably to heavy-handed) or a proper bijective datastructure, such as a perfect hash automaton. The poor man's alternative would be a lexicographically sorted list of words, though that would require a binary search, whereas a perfect hash automaton is linear in the number of bytes/characters).

Of course, you could also implement something like `LinkedHashMap` in Rust with `unsafe`, which would be a better solution, since it does not require two members where one is dependent on the memory of the other. There is in fact such a crate, but it is largely abandoned.

At any rate, before making any savings, I am not convinced that it is really worth it. Word embeddings are typically 300 dimensions * 4 bytes = 1200 bytes. The average heap-allocated token/n-gram is almost two orders of magnitude smaller. It would only make much sense for quantized embeddings, where each embedding (with 300 dimensional embeddings) would be 150 bytes. But then you'd probably save quite a bit more through prefix/suffix sharing of dictionary automata than this type of string sharing.

Reply:
> It does cause issues. Clone the vocabulary, let the original go out of scope, your cloned vocab has dead pointers (to the original Strings. So, you have to implement Clone yourself. And that's the problem, once you remove this safety, with every API change you have to take into account whether there is a situation where it could result in wild pointers. These are exactly the kinds of issues that the Rust ownership system prevents.

Thanks for the explanation, I didn't take cloning into account. That is the reason why I made this draft PR, to make sure that I wasn't missing any scenarios.

> If we want the memory savings (for regular vocabularies it's not so pressing), use proper sharing (Rc, but we cannot do that since it is not Send + Sync, and Arc is probably to heavy-handed) or a proper bijective datastructure, such as a perfect hash automaton. The poor man's alternative would be a lexicographically sorted list of words, though that would require a binary search, whereas a perfect hash automaton is linear in the number of bytes/characters).

I was looking into hash-automata before but abandoned it for some reason that I forgot by now. Iirc, https://docs.rs/fst/0.3.5/fst/ required insertion in lexicographic order, so I guess the index wouldn't be frequency sorted anymore.

> Of course, you could also implement something like LinkedHashMap in Rust with unsafe, which would be a better solution, since it does not require two members where one is dependent on the memory of the other. There is in fact such a crate, but it is largely abandoned.

https://github.com/bluss/indexmap looks maintained and seems to offer a list-map hybrid. But I think any of these solutions mean a change to the API since we can't offer a view into the list as `&[String]` since the values probably aren't placed in a dense vector.

> At any rate, before making any savings, I am not convinced that it is really worth it. Word embeddings are typically 300 dimensions * 4 bytes = 1200 bytes. The average heap-allocated token/n-gram is almost two orders of magnitude smaller. It would only make much sense for quantized embeddings, where each embedding (with 300 dimensional embeddings) would be 150 bytes. But then you'd probably save quite a bit more through prefix/suffix sharing of dictionary automata than this type of string sharing.

I know that the gains would be rather modest, but it looked like a nice opportunity to cut a bit of the memory requirements.

Reply:
> I was looking into hash-automata before but abandoned it for some reason that I forgot by now. Iirc, https://docs.rs/fst/0.3.5/fst/ required insertion in lexicographic order, so I guess the index wouldn't be frequency sorted anymore.

IIRC `fst` only gives you a string -> int mapping and not an int -> string mapping, because it is not a perfect hash automaton (it stores the actual value). In contrast to e.g. my Java [Dictomaton](https://github.com/danieldk/dictomaton) library, which stores right-language cardinalities in states or transitions, which allows you to go from numbers to strings.

A downside would indeed be that strings are sorted in lexicographical order, so you either need to reshuffle the rows (expensive) or add some indirections (which adds more memory again, probably not worth it). This was one of the reasons my initial plans was to also support some on-disk perfect hash automaton format. But I put that on hold because it puts a lot of burden on someone that wants to implement finalfusion in another language.

> https://github.com/bluss/indexmap looks maintained and seems to offer a list-map hybrid. But I think any of these solutions mean a change to the API since we can't offer a view into the list as `&[String]` since the values probably aren't placed in a dense vector.

If it's made by bluss, it's acceptable ;). But indeed, we can't do all of these without API changes.

I see why you want to tackle low-hanging fruit. But I think more can be gained by reducing the size of the embedding matrices. Besides quantization and pruning, I guess it would be interesting to look how half-precision matrices perform, since they'd save half of the memory in non-quantized embeddings. Of course, in the API we'd still need to expose `f32`.